emu/src/cpu/instruction.rs

```diff
@@ -3,35 +3,113 @@ use std::fmt::{Display, Formatter};
 use logger::log;
 
 use crate::bitwise::Bits;
+use crate::cpu::alu_instruction::ArmModeAluInstruction;
+use crate::cpu::arm7tdmi::{Arm7tdmi, REG_PROGRAM_COUNTER};
+use crate::cpu::flags::OperandKind;
+
+use super::condition::Condition;
 
 #[derive(Debug, PartialEq, Eq)]
 pub enum ArmModeInstruction {
-    DataProcessing,
+    DataProcessing {
+        condition: Condition,
+        alu_instruction: ArmModeAluInstruction,
+        set_conditions: bool,
+        op1: u32,
+        op2: u32,
+        destination: u32,
+    },
     Multiply,
     MultiplyLong,
     SingleDataSwap,
-    BranchAndExchange,
+    BranchAndExchange(Condition, usize),
     HalfwordDataTransferRegisterOffset,
     HalfwordDataTransferImmediateOffset,
     SingleDataTransfer,
     Undefined,
     BlockDataTransfer,
-    Branch,
+    Branch(Condition, bool, u32),
     CoprocessorDataTransfer,
     CoprocessorDataOperation,
     CoprocessorRegisterTrasfer,
     SoftwareInterrupt,
 }
 
-impl From<u32> for ArmModeInstruction {
-    fn from(op_code: u32) -> Self {
+impl ArmModeInstruction {
+    pub(crate) fn disassembler(&self) -> String {
+        match self {
+            Self::DataProcessing {
+                condition,
+                alu_instruction,
+                set_conditions,
+                op1,
+                op2,
+                destination,
+            } => {
+                let set_string = if *set_conditions { "s" } else { "" };
+                match alu_instruction {
+                    ArmModeAluInstruction::And
+                    | ArmModeAluInstruction::Eor
+                    | ArmModeAluInstruction::Sub
+                    | ArmModeAluInstruction::Rsb
+                    | ArmModeAluInstruction::Add
+                    | ArmModeAluInstruction::Adc
+                    | ArmModeAluInstruction::Sbc
+                    | ArmModeAluInstruction::Rsc
+                    | ArmModeAluInstruction::Orr
+                    | ArmModeAluInstruction::Bic => {
+                        format!(
+                            "{alu_instruction}{set_string}{condition} R{destination}, 0x{op1:08X}, 0x{op2:08X}"
+                        )
+                    }
+                    ArmModeAluInstruction::Tst
+                    | ArmModeAluInstruction::Teq
+                    | ArmModeAluInstruction::Cmp
+                    | ArmModeAluInstruction::Cmn => {
+                        format!("{alu_instruction}{set_string}{condition} 0x{op1:08X}, 0x{op2:08X}")
```

Comment:
There's not `s` in the ASM syntax for these instructions since they always have the bit `S` set.

And I think we should not display `op1` but `Rn` instead